File: pset6/credit/credit.py
from cs50 import get_int


def main():
    card = str(get_card())

    if check_card(card):
        return print(get_card_types(card))
    return "INVALID"




# function
def get_card():
    while True:
        n = get_int("Number: ")
        if len(str(n)) > 9:
            break
    return n


def check_card(card):
    sum = 0
    for n in range(len(card) - 2, -1, -2):
        digit = int(card[n]) * 2
        if digit > 9:
            digit -= 9
        sum += digit
    for n in range(len(card) - 1, -1, -2):
        sum += int(card[n])
    return True if sum % 10 == 0 else print("INVALID")


def get_card_types(card):
    first_two_digits = int(card[:2])

    if first_two_digits in range(34, 38):
        return "AMEX"
    if first_two_digits in range(51, 56):
        return "MASTERCARD"
    if first_two_digits // 10 == 4:
        return "VISA"
    else:
        return "INVALID"


main()

File: pset1/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int lunhCheckCard();
int countDigitCard();
int selectCardDigit();

int main(void)
{
    long input;
    do
    {
        input = get_long("card: ");
    }
    while (countDigitCard(input) <= 10);
        return lunhCheckCard(input);
}

int countDigitCard(long number) // retorna a quantidade de digítos do cartão
{
    int count = 1;
    while (number > 0)
    {
        number /= 10;
        count++;
    }
    return count;
}

int selectCardDigit(long card, int index) // retorna os dois primeiros digítos do cartão
{
    int size = countDigitCard(card) - index;

    int count = 1;
    while (count < size)
    {
        card /= 10;
        count++;
    }
    return card;
}

int lunhCheckCard(long card) // algoritimo de validação de cartaoes de credito
{
    int oddPosition = 1;
    int evenPosition = 10;
    int sumPair = 0;
    int sumOdd = 0;

    long control = card;
    while (control > 0)
    {
        // multiplica todos os digítos pares e a soma deles
        int pair = (control / evenPosition) % 10;
        pair *= 2;
        if (pair > 9)
        {
            pair -= 9;
        }
        sumPair += pair;

        // soma todos os digítos ímpares
        int odd = (control / oddPosition) % 10;
        sumOdd += odd;

        // lógica
        evenPosition *= 10;
        oddPosition *= 10;
        control /= 10;
    }

    // informações do cartão
    int sumTotal = sumPair + sumOdd;
    int sizeCard = countDigitCard(card);
    int firstDigit = selectCardDigit(card, 1);
    int firstTwoDigits = selectCardDigit(card, 2);

    // verifica se o cartão é valído
    bool VALID = sumTotal % 10 == 0 && sizeCard >= 13;
    if (VALID)
    {
        // verifica as bandeiras dos cartões
        bool isAMEX = firstTwoDigits == 34 || firstTwoDigits == 37;
        bool isMASTERCARD = firstTwoDigits >= 50 && firstTwoDigits <= 55;
        bool isVISA = firstDigit == 4;
        if (isAMEX && VALID)
        {
            printf("AMEX\n");
        }
        else if (isMASTERCARD && VALID)
        {
            printf("MASTERCARD\n");
        }
        else if (isVISA && VALID)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}
